src/collectors/proc.plugin/sys_class_power_supply.c

```diff
@@ -116,14 +130,83 @@ static void add_labels_to_power_supply(struct power_supply *ps, RRDSET *st) {
     rrdlabels_add(st->rrdlabels, "device", ps->name, RRDLABEL_SRC_AUTO);
 }
 
+static bool read_simple_property(struct simple_property *prop, bool keep_fds_open) {
+    char buffer[30 + 1];
+
+    if(unlikely(prop->fd == -1)) {
+        prop->fd = open(prop->filename, O_RDONLY | O_CLOEXEC, 0666);
+        if(unlikely(prop->fd == -1)) {
+            collector_error("Cannot open file '%s'", prop->filename);
+            return false;
+        }
+    }
+
+    if (likely(prop->fd != -1))
+    {
+        ssize_t r = read(prop->fd, buffer, 30);
+        if(unlikely(r < 1)) {
+            collector_error("Cannot read file '%s'", prop->filename);
+            close(prop->fd);
+            prop->fd = -1;
+            return false;
+        }
+        else {
+            buffer[r] = '\0';
+            prop->value = str2ull(buffer, NULL);
+        }
+
+        if(unlikely(!keep_fds_open)) {
+            close(prop->fd);
+            prop->fd = -1;
+        }
+        else if(unlikely(lseek(prop->fd, 0, SEEK_SET) == -1)) {
+            collector_error("Cannot seek in file '%s'", prop->filename);
+            close(prop->fd);
+            prop->fd = -1;
+        }
+    }
```

Comment:
Considering that you are having:

```sh
 close(prop->fd);
prop->fd = -1;
```

on three difference places, and on two places you are also calling `collector_error`, probably you can rework this part of the code to have as an inline function.

Reply:
I did not use inline function, but simplified error handling in the way similar to how it's done for multi-dimensional properties